### PR TITLE
use aes-192-cbc as encryption

### LINKED ISSUE #660: Replace encrypt/decrypt with a real encryption like AES-256
See discussion starting at https://github.com/ioBroker/ioBroker.js-controller/issues/262#issuecomment-584343117

### PR DESCRIPTION
**Pull request by foxriver76**
- backward compatible with current encryption
- closes #660

### CODE REVIEW

File: lib/tools.js [hunk 1]
```diff
@@ -1847,14 +1847,65 @@ function encrypt(key, value) {
...
+ */
+function encrypt(key, value) {
+    if (key.length < 32) {
+        // key is too short for AES-256-CBC - fallback to old encryption
+        return encryptLegacy(key, value);
+    }
+
+    crypto = crypto || require('crypto');
+
+    key = key.substr(0, 32);
```
Comment (Apollon77) [inline lib/tools.js:1873]: means that code that the key needs to be exact 32 chars ... or only first 32 chars are used?

**Comment by Apollon77** [comment]
LGTM ...
Questions:
1. Admin also needs that code, correct?! Or something special?
2. Additionally we need to generalte a new type of keys for all new installations because it will not be compatible or?! and so only new installations will use the new encryption ... ?

File: lib/tools.js (hunk 1 above)
Comment (foxriver76) [reply lib/tools.js:1873]: We only use first 32 chars of the secret, because secret is too long.

**Comment by foxriver76** [comment]
1. Admin basically needs the same code, but here we are using nodejs crypto module. Admin of course needs another lib.

2. No, the keys that I know are longer than 32 chars, so its compatible, only our fallback secret is too short

File: lib/tools.js (hunk 1 above)
Comment (AlCalzone) [reply lib/tools.js:1873]: How long is the secret and which characters does it contain? 32 ascii chars seems a bit weak for a 256-bit key.

Comment (foxriver76) [reply lib/tools.js:1873]: You are right. I will look into it later.

Comment (foxriver76) [reply lib/tools.js]: secrets are generated with 192bits in https://github.com/ioBroker/ioBroker.js-controller/blob/master/lib/setup/setupSetup.js#L190

Thus, I changed encryption to aes 192 cbc so we can use full secret. Changing the secret will most likely be breaking and 192 bit should be fine.

**Comment by foxriver76** [comment]
Ready to merge from my pov

File: lib/tools.js [hunk 2]
```diff
@@ -1847,14 +1847,63 @@ function encrypt(key, value) {
...
+ * @param {string} key - Secret key
+ * @param {string} value - value to decrypt
+ * @returns {string}
+ */
+function encrypt(key, value) {
+    if (key.length !== 48) {
+        // key is not matching for AES-192-CBC - fallback to old encryption
+        return encryptLegacy(key, value);
+    }
+
```
Comment (AlCalzone) [inline lib/tools.js:1870]: This should also apply if key is not a valid hex string

File: lib/tools.js [hunk 3]
```diff
@@ -1847,14 +1847,63 @@ function encrypt(key, value) {
...
+ * encrypts a value by a given key via AES-192-CBC
+ *
+ * @param {string} key - Secret key
+ * @param {string} value - value to decrypt
+ * @returns {string}
+ */
+function decrypt(key, value) {
+    if (!value.startsWith(`$/aes-192-cbc:`) || key.length !== 48) {
+        return decryptLegacy(key, value);
+    }
```
Comment (AlCalzone) [inline lib/tools.js:1892]: dito

File: lib/tools.js [hunk 4]
```diff
@@ -1847,14 +1847,63 @@ function encrypt(key, value) {
...
+        return encryptLegacy(key, value);
+    }
+
+    crypto = crypto || require('crypto');
+
+    const iv = crypto.randomBytes(16);
+    const cipher = crypto.createCipheriv('aes-192-cbc', Buffer.from(key, 'hex'), iv);
+    let encrypted = cipher.update(value);
+
+    encrypted = Buffer.concat([encrypted, cipher.final()]);
```
Comment (AlCalzone) [inline lib/tools.js:1877]: If you're not planning to do anything else with `encrypted`, I'd do
```js
const encrypted = Buffer.concat([cipher.update(value), cipher.final()]);
```

File: lib/tools.js [hunk 5]
```diff
@@ -1847,14 +1847,63 @@ function encrypt(key, value) {
...
+
+    crypto = crypto || require('crypto');
+
+    const textParts = value.split(`:`);
+    const iv = Buffer.from(textParts[1], 'hex');
+    const encryptedText = Buffer.from(textParts.pop(), 'hex');
+    const decipher = crypto.createDecipheriv('aes-192-cbc', Buffer.from(key, 'hex'), iv);
+    let decrypted = decipher.update(encryptedText);
+
+    decrypted = Buffer.concat([decrypted, decipher.final()]);
```
Comment (AlCalzone) [inline lib/tools.js:1902]: see comment regarding `encrypted`

**Comment by klein0r** [comment]
Exists a documentation how to use that in own adapters and what are the requirements/dependencies to encrypt data this way?

**Comment by foxriver76** [comment]
Yes, we refer to https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/adaptersecurity.md

Note, that the `aes-192-cbc` algorithm will be used up from js-controller v3.2 - until then we use our 'old' encryption algorithm. 

Additionally (js-c >= 3.1) you can use 
```javascript 
const encrypted = adapter.encrypt('top secret message');
```
 and

```javascript
const decrypted = adapter.decrypt('encryptedMessage');
```
to use the current encryption algorithm for active usage in your code. Also note, that the underlying algorithm will change with js-c 3.2, but it will stay backward compatible and detect the used encryption algorithm and thus applies the correct decryption technique.

**Pull request merged**